### PR TITLE
Mainloop multistaged pipeline and instruction scheduling for NVIDIA Ampere Tensor Cores (F16 and F32)

### PR DESCRIPTION
**Pull request by manishucsd**
This PR creates a coarse-grained multistage pipelining and fine-grained instruction scheduling for optimal performance on NVIDIA Ampere Tensor Cores.

Multi-staging is essential to hide the Global Memory load latency by building longer software pipelines and using the available Shared Memory capacity, especially on NVIDIA A100. Additionally, fine-grained instruction scheduling hides the Shared Memory load latency by prefetching the math operands into registers. The PR adds support for F16 and F32 datatype using `mma.sync` native Tensor Core exposure in the nvptx.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Replace with:
```
#define DEBUG_TYPE "iree-gpu-pipelining"
````
as this is common to the whole file

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir]: nit: remove empty line

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir:148]: Are we missing some checks for the B matrix loads?

File: compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir:494]: Can we add an equivalent test for f32 type?

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: nit: This doesn't add much information, I would remove it.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: ditto: I would remove

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:324]: This is not safe as it assumes the op has an operand without know what it is.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Remove useless comments.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: nit, use this form: `/*kgroup=*/0`

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:414]: nit: Can you add a comment on why we need to scan for dependency after having done it for async.cp ops. I always lose track of that part.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: mark this function as static.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:484]: we should be able to support `numStages == 2`? Would the code not work?

**Review by ThomasRaoux — COMMENTED** [review]
Looks good overall. Mostly nits comments. I'm curious why we have to restrict to numStages >= 3?

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Done

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Done

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Done

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Done

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:484]: https://github.com/openxla/iree/pull/12550#discussion_r1129931689

> We need 3 or more stages for Ampere-style multi-staged pipeline: 0th stage for `cp.async`, last stage for `mma.sync` and some `ld.shared`/`ldmatrix`, one before the last stage for load to first kgroup. 

We schedule in stage 0, numStages - 1, and numStages - 2. If this condition is not met, then ops for 0th stage and numStages - 2 goes to the same stage instead of two different stage. I don't know what effect it has on functionality and performance. We can try it out when and if we need it. However, on any Ampere architecture, it won't make sense to have lower than 3 stages.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:414]: If I switch the order, I get the following error that we have also seen when we fixed it. 

```
iree-compile: /mnt/disks/gcloud_workspace/repos/iree/iree_tree_4/iree/third_party/llvm-project/mlir/lib/Dialect/SCF/Transforms/LoopPipelining.cpp:229: auto (anonymous namespace)::LoopPipelinerInternal::analyzeCrossStageValues()::(anonymous class)::operator()(mlir::OpOperand &) const: Assertion `stage > defStage->second' failed.
```

I put in some comment that makes sense to me. Let me know if you have better take or comment here after this debug error information.

File: compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir:494]: Added.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir:148]: Added the checks for `memref.loads` lowered to `llvm.extractvalue`

```
//  CHECK-COUNT-4:   llvm.extractvalue{{.*}} : !llvm.struct<(i32, i32, i32, i32)> 
```

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:324]: Check the checks in the new version :)

**Review by ThomasRaoux — APPROVED** [review]

**Comment by hanhanW** [comment]
I think the policy is not changing the submodule if possible. We usually wait until integrate happening.

Also, is the commit checked in upstream? https://github.com/iree-org/iree-llvm-fork/commit/4e0d4ec3bdc8d6dbb3e6a4c6f4b6dbe7d51e94a5

**Comment by ThomasRaoux** [comment]
> I think the policy is not changing the submodule if possible. We usually wait until integrate happening.
> 
> Also, is the commit checked in upstream? [iree-org/iree-llvm-fork@4e0d4ec](https://github.com/iree-org/iree-llvm-fork/commit/4e0d4ec3bdc8d6dbb3e6a4c6f4b6dbe7d51e94a5)

Yes since Manish is integrator this week and this will be part of the next integrate I believe it is okay. Looks like the patch that landed is slightly different, but the patch has definitely landed.

**Comment by hanhanW** [comment]
> Yes since Manish is integrator this week and this will be part of the next integrate I believe it is okay. Looks like the patch that landed is slightly different, but the patch has definitely landed.

I think we still prefer not doing this even Manish is the integrator this week. There are two reasons that I can think of.

1. Integrate w/o any cherry-picks is cleaner. It makes Google internal integrate much easier. 
2. We don't know when the patch will be propagated. We usually pin the LLVM version to what tensorflow uses. It might be blocked by hard commits. E.g., the dialects were split into many dialects, many hard work were needed. In this context, we could have to cherry-pick the commit for few times, which could lead to a situation that we have to transfer the cherry-pick commits information to the next integrator.

**Comment by powderluv** [comment]
I think there is a subtle but important policy change with the no-cherry-picks. 

It may make the Google integration easier and possibly saves some time for integrations (which is very complex and time consuming so thank you all for doing that) but it changes the open source dynamics of downstream projects. Right now we we have been operating on everything upstream first principle because we know if we fix it in the right spot usually we get someone to help cherry-pick the LLVM fix locally and it is low friction. 

This new policy adds a burden to downstream integrations to carry patches locally in our own IREE-LLVM-Fork (or a branch of IREE-LLVM-Fork) until the next integrate lands. This may be ok but then if we find an issue in SHARK-runtime it is different from the IREE-LLVM-Fork so our bug reporting and debugging is not happening on IREE but in a downstream fork and that diverges over time and we have to do heavy integrates with IREE (today we rebase on IREE Top of Master every two hours).  Case in point https://github.com/openxla/iree/issues/12634 required an LLVM fix that would have shipped to our beta testers by now on the IREE-LLVM-Fork (it fails to compile for the gpu wmma vulkan path). We have created a fork to build with that patch but that is just making it harder to make it upstream first. 

We can manage with the new policy but if someone else reported an issue with stock IREE that required the same fix above they will have to wait a few days to fix the breakage. So please do consider downstream and OSS users as much as the integration into Google.  

fyi @stellaraccident  / @benvanik

**Comment by ThomasRaoux** [comment]
> > Yes since Manish is integrator this week and this will be part of the next integrate I believe it is okay. Looks like the patch that landed is slightly different, but the patch has definitely landed.
> 
> I think we still prefer not doing this even Manish is the integrator this week. There are two reasons that I can think of.
> 
> 1. Integrate w/o any cherry-picks is cleaner. It makes Google internal integrate much easier.
> 2. We don't know when the patch will be propagated. We usually pin the LLVM version to what tensorflow uses. It might be blocked by hard commits. E.g., the dialects were split into many dialects, many hard work were needed. In this context, we could have to cherry-pick the commit for few times, which could lead to a situation that we have to transfer the cherry-pick commits information to the next integrator.

I agree with you on principle. In practice:
1) there was already cherry-picks in the submodule so we were not on a clean LLVM commit.
2) The patch had already been integrated in TF/mhlo (https://github.com/llvm/llvm-project/commit/117db47d02c174e2ec039fa8b6a97381106e6238)

I have been treating the rule about cherry-pick as something where we can use some flexibility when it makes sense. It sounds like there are different interpretations, it could be something we discuss further as a team however I think we have to make sure we don't make things harder for developers than they have to be.

**Comment by stellaraccident** [comment]
Just scanning, I think we are conflating two things:

1. As a general rule, we should be very careful about doing interlocked at-head development between llvm and iree unless if the feature is protected by a flag or something. Basically mainline should be stable and move forward predictably without regressing based on upstream changes. This may not always be possible (either because truly impossible or because we decided on a case that it is not worth it and are going to violate our general rule). In this case, we must move then together in lock step so that main does not regress or exist in a "half implemented" state waiting for an eventual version bump to resolve. Today that means that we cherrypick, even if as a general rule, we don't like doing that as part of new development.
2. Cherrypicks are expensive for the whole ecosystem because they reduce the frequency of "milestone commits" where we have a clean dep on upstream. It is a mark of maturity of a project like this that it rarely needs them and then typically only uses them for bug fixes and critical hot fixes. Using them continuously for ongoing interlocked development is a sign of unhealthy coupling, generally.

So, in short, the need to do interlocked cherrypicks to move mainline forward is a sign of suboptimality in our dev process. But if we're in that state, it is something we should aim to do better next time -- but we need to do whatever work is necessary to ensure that mainline moves forward and does not regress. If this means carrying patches, using smoke signals, etc, that is what we do. The bug to be fixed for the next time in such cases is not the cherrypick itself but the sequence of development work that resulted in an interlocked state.

**Comment by stellaraccident** [comment]
I heard that not doing the cherrypick in this case has resulted in a regression in mainline? Is that true? If so, we have to resolve that, and waiting for an integrate isn't the right answer.

**Comment by stellaraccident** [comment]
Note that i think Anush's analysis was more appropriately targeted at https://github.com/openxla/iree/issues/12634, which was a fix for a field breakage. This one is new development and Hanhan is correct that we do try to avoid interlocked at head development. But since already down that path, keep it in mind for later.

**Pull request merged**
